src/tir/transforms/split_host_device.cc

```diff
@@ -89,6 +92,17 @@ class VarUseDefAnalysis : public StmtExprMutator {
 
   Stmt VisitStmt_(const AllocateNode* op) final {
     this->HandleDef(op->buffer_var.get());
+    auto storage_scope = runtime::StorageScope::Create(GetPtrStorageScope(op->buffer_var));
+    if (storage_scope.rank == runtime::StorageRank::kDynShared) {
+      ICHECK_EQ(use_dyn_shmem_, false) << "Only one dynamic shared memory allocation is allowed.";
```

Comment:
Shall we support allocating multiple arrays from dynamically allocated shared memory in the future?
```cuda
extern __shared__ float buffer[];
float* arr1 = (float*)buffer;
int* arr2 = (int*)&arr1[128];
...
```

Reply:
I'm planning to support allocating multiple dynamic shmem at TIR level, like:
```
A_shared = ib.allocate(dtype, A.shape, scope="dyn.shared")
B_shared = ib.allocate(dtype, B.shape, scope="dyn.shared")
``` 
Under the hood, multiple allocations will be merged into a single allocation via `storage_rewrite.cc` pass. But I haven't thought about merging multiple allocations with hetero-dtype like your example. 

Maybe for dynamic shmem, we just allocate a byte array and use `reinterpret` cast for splitting, assuming it is possible with TIR?  Probably `storage_rewrite.cc` does not work for such scenario. Thought? @tqchen @junrushao1994 @vinx13 

Also @yzh119 do you have a use case for such hetero-dtype dynamic shared memory?

Reply:
@masahi this happens when you need to load both data and index into shared memory (e.g. sparse matrix multiplications).

Reply:
We can do a lowering stage that translate allocate dynamic memory to an intrinsic that directly get the dynamic memory with an offset(where multiple allocation can be merged with proper alignment). This will help us get around some limitation imposed by storage rewrite. Split host device might be the right place for this.

Reply:
I realized that we must support this usage, since it is required, among others, for multiplying fp16 or int8 matrices and accumulate into fp32 or int32.

I'm thinking that we need to expose pointer cast in TIR. I dind't find it in the current code:
```
extern __shared__ char buf[SIZE];
half* A_ptr = (half*)&buf[0];
half* B_ptr = (half*)&buf[128];
float32* C_ptr = (float32*)&buf[256];
```
I didn't get what you meant by "translate to an intrinsic" @tqchen. Can we hide pointer cast and the load /stores that follow it behind intrinsics? For allocation and merging, I'm thinking about working exclusively with byte arrays and make use of `storage_rewrite` as is.

Reply:
One possible solution is to rewrite the above program to 
```
extern __shared__ char buf[SIZE];
*((floa32*)&buf[256]) = ((half*)&buf[0])[0] * ((half*)&buf[128])[0];
```
And introduce two intrinsics, `reinterpret_load` / `reinterpret_store`, which can be used to express the snippet above by
```
extern __shared__ char buf[SIZE];
reinterpret_store(&buf[256], "float32", 0, 
 reinterpret_load(&buf[0], "half", 0) * reinterpret_load(&buf[128], "half", 0));
```